Allow configuring static IP and MAC with primary UDNs.

qinqon opened the pull request:
## 📑 Description
Allow a pod's primary Layer 2 UDN network to select static IP and MAC addresses using the Multus default-network annotation.

Fixes https://issues.redhat.com/browse/CORENET-6131

## Additional Information for reviewers
Depends-on:
- https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5320

Notes:
- Collision detection is not included in this PR and will be addressed in a subsequent update.
  - https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5411 
- No additional checks will be added to the CNI, as CNI code respects OVN pod annotations and disregards IP requests.
- Negative test cases for pods will be incorporated by the QE team in a follow-up PR.
- NAD client was bumped but we needed to ping Pin containernetworking/cni v1.2.3 at go.mod looks like future versions will break ovnk

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [x] My code requires tests
- [x] if so, I have added and/or updated the tests as required
- [x] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
This include unit test and e2e tests

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added support for static IP and MAC address assignment on user-defined network interfaces for pods and KubeVirt VMs.
  * Enhanced network status reporting to include MTU information.
  * Introduced new end-to-end tests covering static IP/MAC assignment and validation for both pods and VMs.
  * Improved feature flag and environment variable handling for enabling preconfigured UDN addresses.
  * Extended CNI plugin support with new commands (GC, STATUS) and improved error handling.
  * Added network namespace support for multiple OS platforms with platform-specific implementations.
  * Refined network selection logic to incorporate default network annotations when feature enabled.
  * Added utility functions for filtering IPs by cluster IP family support and generating address annotations.

* **Bug Fixes**
  * Improved validation and error reporting for static IP requests in certain network scenarios.

* **Dependency Updates**
  * Upgraded CNI and network-attachment-definition-client dependencies to latest versions.

* **General Improvements**
  * Enhanced test coverage for custom network configurations and static address scenarios.
  * Modernized file I/O and version handling in CNI libraries.
  * Improved utility functions for IP family detection and annotation generation.
  * Refined CNI plugin initialization and configuration parsing with support for multiple CNI versions and flags.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

maiqueb commented on contrib/kind-common:
```diff
@@ -434,6 +434,16 @@ install_multus() {
     sed -e "s|multus-cni:snapshot|multus-cni:${version}|g" |\
     run_kubectl apply -f -
+
+    run_kubectl apply -f - <<EOF
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: default
+  namespace: ovn-kubernetes
+spec:
+  config: '{"cniVersion": "0.4.0", "name": "ovn-kubernetes", "type": "ovn-k8s-cni-overlay"}'
+EOF
```
maybe you could rebase this on top of #5320 and avoid having to use this ?

maiqueb replied:
Tks; this way, you help me test my stuff :stuck_out_tongue_closed_eyes:

maiqueb commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1344,18 +1344,31 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *corev1.Pod, nInfo NetInfo,
+
+	// If there are static IPs and MACs at the default NSE, override the active NSE with them
+	// TODO: Discover default network "name" and "namespace"
+	if defaultNSE != nil && defaultNSE.Namespace == "ovn-kubernetes" && defaultNSE.Name == "default" {
```
I think you can hardcode "default" NAD metadata name, and can rely on `config.Kubernetes.OVNConfigNamespace` to get the namespace.

qinqon replied:
You have being a living LLM to ask for :-)

qinqon commented:
Looks like bumping net-attach-def client breaks networkmanage unit tests

qinqon commented:
when an existing NAD is passed to the default network then it replace pod request nad name 

```
```ovnk ovn-worker2 logs ovnkube-node ovnkube-controller |grep DELETE
I0701 11:17:54.410482   61361 cniserver.go:131] DELETEME, cr.Config: %s{"cniVersion":"1.0.0","name":"default-network-annotation-7332_secondaryl2network","netAttachDefName":"default-network-annotation-7332/secondaryl2network","role":"secondary","subnets":"103.0.0.0/16,2014:100:200::0/60","topology":"layer2","type":"ovn-k8s-cni-overlay"}

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation_test.go:
```diff
@@ -871,6 +1009,15 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
 			)
 
+			if tt.wantErr {
+				// check the expected error after we have checked above that the
+				// rollback has behaved as expected
```
This should not go here, this should be at the place it was I will move back after rollback check.

qinqon commented on go-controller/pkg/clustermanager/defaultnetworkcustomization/ensurenad.go:
```diff
@@ -4,13 +4,13 @@ import (
 	nadclientset "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned"
 	nadlisters "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/listers/k8s.cni.cncf.io/v1"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
will check

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -330,13 +353,11 @@ func allocatePodAnnotationWithRollback(
-		// be allocated or was already allocated by the same pod
-		err = fmt.Errorf("cannot allocate a static IP request with IPAM for pod %s", podDesc)
-		return
+		if err = validateStaticIPRequest(netInfo, podDesc); err != nil {
```
> Just to make sure we are all on the same page, this is all we need to ensure that the requested IP ends up on the IPAMClaim too right?

Yes, with this IPAMClaims status store the IP from IPRequest

> Is the scenario when someone requests a different IP than the one that's set already on the existing IPAMClaim covered as well?

I understand we don't support ipamclaims for pods so users cannot creat a new pod pointing to a exiting ipamclaim with ip at status, and for VMs addresses are inmutable and users cannot reference ipamclaims from VMs.

maiqueb replied:
This is an interesting train of thought. Tks @kyrtapz for making me think this through.

Looking at the code below, I see the following code:
```golang
	     if hasIPRequest {
			tentative.IPs, err = util.ParseIPNets(network.IPRequest)
			if err != nil {
				klog.Warningf("Failed parsing IPRequest %+v for pod %s: %v", network.IPRequest, podDesc, err)
				return
			}
		} else if hasIPAMClaim {
			tentative.IPs, err = util.ParseIPNets(ipamClaim.Status.IPs)
			if err != nil {
				return
			}
		}
```

Which means - IIUC - that:
- the IP request will always take precedence to the IPAMClaim IP. 
- the IP requests cannot change on an existing pod - there's (or there will be) the Kubernetes VAP to prevent it
- thus, the only way we can find ourselves in the situation described - IPRequests != IPAMClaim.Status.IPs - is if the IPAMClaim IPs change. 

How can that happen ? 
I'm trusting a mix of RBAC - the component that *owns* the status is OVN-Kubernetes - and proper test coverage to ensure that doesn't happen. 

Maybe we could add some CEL validation (please bear with me, maybe this suggestion is not possible) to the IPAMClaim status to prevent the status IPs from changing ? It would need to accept the *first* status IPs definition, but from then, keep in immutable. How could we express that ?

ormergi replied:
The following rule should  enforce mutability: 
```
	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Status is immutable"
```

ormergi replied:
Making the status immutable sounds odd and some what out of the convention where it should refect the current state, semantically if its immutable it will never reflect a current state :sweat_smile: 

I need to go over this PR to suggest something meaningfull

maiqueb replied:
will this allow the controller to set the state initially ? i.e. define the IPs once ? 

Keep in mind we first (ipam-ext) creates the IPAMClaim, then a separate component - OVN-K - mutates this ipam claim, filling out its status attribute.

ormergi replied:
I think the following rule should work as long the IPAMClaim CR status.IPs is empty once created and remain empty until OVN-K updates it:
```
// +kubebuilder:validation:XValidation:rule="self == '''' ||  self == oldSelf", message="status.IPs cannot be changed once set"
```

Hardening the status up to it can be updated one time only feel wrong to me, by the convention status should reflect the current state, having it immutable (eventually) it cannot reflect a current state - semantically.

The way I see it, when IP-request mismatch the referenced IPAMClaim spec, its an undefined state caused by some component that did wrong. OVN-K should blow up in that case and return an error. 
Do you think this can work?

ormergi commented:
@qinqon #5332 is merged could you please rebase and drop the first commit?

ormergi reviewed (commented):
Very nice!
This is first round, I need to go over e2e again, please see my comments

maiqueb commented on go-controller/pkg/util/multi_network_test.go:
```diff
@@ -1048,11 +1166,14 @@ func TestGetPodNADToNetworkMappingWithActiveNetwork(t *testing.T) {
 
-			if err != nil {
+			if test.expectedError != nil {
+				g.Expect(err).To(gomega.HaveOccurred(), "unexpected success operation, epecting error")
 				g.Expect(err).To(gomega.MatchError(test.expectedError))
```
aren't these redundant ? i.e. I'd just leave the expectation with the `MatchError` matcher.

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation_test.go:
```diff
@@ -575,6 +580,130 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
+			name: "expect error, static IP with IPAM on layer2 when EnablePreconfiguredUDNAddresses is false",
+			ipam: true,
+			// enablePreconfiguredUDNAddresses defaults to false
+			netConf: &ovncnitypes.NetConf{
```
good point.

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -207,6 +207,28 @@ func allocatePodAnnotationWithTunnelID(
 }
 
+// validateStaticIPRequest checks if a static IP request can be honored when IPAM is enabled for the given network.
+func validateStaticIPRequest(netInfo util.NetInfo, podDesc string) error {
```
@ormergi you lost me at "IPAMClaim CR IP request". AFAIK there isn't an IP request attribute in IPAMClaims.

Do you mean in the status ? If so, it would OK - we'd just have to *also* allow it if `status.ips` are empty. Right ?

qinqon replied:
I will add a validation to check that IPRequest is the same as IPAMClaim.Status.IPs ]

tssurya replied:
I don't see this fixup.. lmk if I missed it somewhere as I had the same question

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation_test.go:
```diff
@@ -575,6 +580,130 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
+			name:                            "expect requested MAC and IP, IPAM",
+			ipam:                            true,
+			enablePreconfiguredUDNAddresses: true,
+			role:                            types.NetworkRolePrimary, // has to be primary network for default routes to be set
+			netConf: &ovncnitypes.NetConf{
+				Topology: types.Layer2Topology,
+				NetConf: cnitypes.NetConf{
+					Name: "network",
+				},
+				NADName: util.GetNADName("namespace", "network"),
+				Subnets: "192.168.0.0/24,2001:db8::/64",
+				Role:    types.NetworkRolePrimary,
```
this is fishy. We're passing the role attribute directly, but that is part of the `NetConf` struct. 

At the very least - IMHO - we can only pass this attribute via one way.

qinqon replied:
@maiqueb I am going to remove the "netConf" option totally from the tests, I was just missing the allowPersistentIPs flag on the test
to make it work

maiqueb replied:
that works for me - we either pass netconf w/ everything in it, *or* we pass the individual knobs and compose a netconf in the test. 

Doing both makes this even harder to follow.

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation_test.go:
```diff
@@ -575,6 +580,130 @@ func Test_allocatePodAnnotationWithRollback(t *testing.T) {
+			wantErr: true,
+		},
+		{
+			// on primary networks with IPAM and layer2 topology, expect success when EnablePreconfiguredUDNAddresses is enabled
```
I am sooooo confused. This has too many parameters, making it hard to follow. Or I'm daft. 

Isn't this the same test you have added on line 585 ?

maiqueb replied:
could we kind of tinker with the test names or comments to reflect that ? 

honestly, I get headaches from looking at these test entries.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -1619,6 +1625,32 @@ write_files:
+					if !isPreConfiguredUdnAddressesEnabled() {
+						Skip("ENABLE_PRE_CONF_UDN_ADDR not configured")
+					}
+					staticIPs, err := json.Marshal(map[string][]string{
+						"net1": filterCIDRs(fr.ClientSet, staticIPv4, staticIPv6),
+					})
+					Expect(err).ToNot(HaveOccurred())
+					annotations := map[string]string{
+						kubevirtAddressesAnnotation: string(staticIPs),
+					}
```
maybe turn this into an helper func with a proper name ?

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -1619,6 +1625,32 @@ write_files:
+						kubevirtv1.NetworkSource{
+							Pod: &kubevirtv1.PodNetwork{},
+						}, userDataWithIperfServer, networkDataDualStack)
+					vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].Bridge = nil
+					vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].Binding = &kubevirtv1.PluginBinding{Name: "l2bridge"}
+					vm.Spec.Template.Spec.Domain.Devices.Interfaces[0].MacAddress = staticMAC
```
nit: maybe just setting this once using an `Interface` struct ? you'd then have a well named helper return that `Interface`, something like: `vm.Spec.Template.Spec.Domain.Devices.Interfaces[0] = primaryUDNIface(withMACAddr(staticMAC))`

you can ignore this comment.

qinqon commented on test/e2e/kubevirt.go:
```diff
@@ -1619,6 +1625,32 @@ write_files:
+				cmd: func() string {
+					GinkgoHelper()
+					if !isPreConfiguredUdnAddressesEnabled() {
+						Skip("ENABLE_PRE_CONF_UDN_ADDR not configured")
```
> Please consider replacing the skip with actual assertion, let the tests runner skip the tests according to env.

@ormergi what do you mean ?

qinqon commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1321,6 +1321,21 @@ func GetPodNADToNetworkMapping(pod *corev1.Pod, nInfo NetInfo) (bool, map[string
 
+// overrideActiveNSEWithDefaultNSE overrides the provided active NetworkSelectionElement with the IP and MAC requests from
+// the default NetworkSelectionElement after validating its namespace and name.
+func overrideActiveNSEWithDefaultNSE(defaultNSE, activeNSE *nettypes.NetworkSelectionElement) error {
```
I don't thing it's worth it, it means cloning active and returning it.

maiqueb commented on go-controller/pkg/clustermanager/routeadvertisements/controller.go:
```diff
@@ -1016,7 +1017,7 @@ func (c *Controller) getSelectedNADs(networkSelectors apitypes.NetworkSelectors)
 			// if we are selecting the default networkdefault network label,
 			// make sure a NAD exists for it
-			nad, err := c.getOrCreateDefaultNetworkNAD()
+			nad, err := defaultnetworkcustomization.EnsureDefaultNetworkNAD(c.nadLister, c.nadClient)
```
@RamLavi please check the file you've linked. It's changed there. 

Are you sure something's wrong ?

kyrtapz reviewed (approved):
LGTM

maiqueb reviewed (dismissed):
tks.

tssurya reviewed (commented):
@qinqon great job! I don't have any major concerns, just some questions for my own understanding.

Also did you add validation for this: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5321/commits/f42a91e9c5b2f44c86c55be7556c0b0f9da0b4dd#r2224681260 if so where?

OK this PR was simple enough, it does the 2 main things:

1. allows IP Request and IPAMClaim to co-exist for L2 Primary UDNs
2. allows the re-use of `v1.multus-cni.io/default-network` as the active NSE for primary UDN L2 pods - if we were to add the annotation on the primary network like we do for secondary - its going to delay pod setup which is also why for P-UDNs we decided to plumb in 2 interfaces together at the same time with 1 OVNK IPAM annotation - so we stick with the same thing but re-use the default network annotation+default NAD instead of re-doing `k8s.v1.cni.cncf.io/networks` annotation - as agreed on in OKEP design for this feature. 

Missing bits:
3. Doesn't do any validation on conflicts with the IPRequest field - blindly uses it @RamLavi 's PR will bring the validation in
4. Doesn't do any validation on conflicts with the MAC field - blindly uses it - needs a FUP.
5. E2E's which @asood-rh should add in FUP

tssurya commented:
@qinqon if you get some time - ensuring when the feature is enabled and the pod network is a secondary network nothing changes for secondary pods would be good to confirm - if there are tests already for this then good.

qinqon commented on go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go:
Right, I will resolve this

qinqon commented on go-controller/cmd/ovn-k8s-cni-overlay/ovn-k8s-cni-overlay.go:
```diff
@@ -20,10 +20,12 @@ func main() {
-			p.CmdAdd,
-			p.CmdCheck,
-			p.CmdDel,
+		skel.PluginMainFuncs(
```
Yep, is this ok ? or you see some issue with it ?

tssurya replied:
no all good, resolving it

qinqon commented on go-controller/go.mod:
If I remember correctly using cni 1.3.0 (latest stable) was breaking ovn-k compilation, but I have try it again and it's compiling allright
maybe something changed at ovn-kubernetes and not it's fine ? 

Long Story Short I am not able to reproduce the compilation error, I will pin "cni" to the latest stable wich is "1.3.0"

tssurya replied:
ack that makes sense ! thanks

qinqon commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1349,18 +1364,39 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *corev1.Pod, nInfo NetInfo,
+	if IsPreconfiguredUDNAddressesEnabled() {
+		// Limit the static ip and mac requests to the layer2 primary UDN when EnablePreconfiguredUDNAddresses is enabled, we
+		// don't need to explicitly check this is primary UDN since
+		// the "active network" concept is exactly that.
```
> when we say we don't need to explicitly check if its primary UDN since active network concept does that,, we mean:
`activeNetwork.TopologyType() == types.Layer2Topology` -> its not like default network is Layer2 and secondary networks active network will be default network or the primary one anyways ?

The activeNetwork is the primary network is not related to secondary networks at all, like a secondary network's primary network can be whatever users wants, for example a pod on layer3 
primary network that happends to also have a secondary layer2 network.

qinqon commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1349,18 +1364,39 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *corev1.Pod, nInfo NetInfo,
+			if err != nil {
+				return false, nil, fmt.Errorf("failed getting default-network annotation for pod %q: %w", pod.Namespace+"/"+pod.Name, err)
+			}
+			// If there are static IPs and MACs at the default NSE, override the active NSE with them
```
Right, what we did here is to extend the logic a little and pass NSE to active network so is no longer always empty.

qinqon commented on go-controller/pkg/util/multi_network_test.go:
```diff
@@ -1011,10 +1012,143 @@ func TestGetPodNADToNetworkMappingWithActiveNetwork(t *testing.T) {
+				DefNetworkAnnotation:         `[{"foo}`,
+			},
+			enablePreconfiguredUDNAddresses: true,
+			expectedError:                   fmt.Errorf(`failed getting default-network annotation for pod "/test-pod": %w`, fmt.Errorf(`GetK8sPodDefaultNetwork: failed to parse CRD object: parsePodNetworkAnnotation: failed to parse pod Network Attachment Selection Annotation JSON format: unexpected end of JSON input`)),
```
Do you mean adding default-network annotation pointing to a secondary network ?

tssurya replied:
yes!

qinqon replied:
The unit test before this one is already doing that I think 
```golang
{
			desc: "default-network ips and mac is is ignored for Layer3 topology",
			inputNetConf: &ovncnitypes.NetConf{
				NetConf:  cnitypes.NetConf{Name: networkName},
				Topology: ovntypes.Layer3Topology,
				NADName:  GetNADName(namespaceName, attachmentName),
				Role:     ovntypes.NetworkRolePrimary,
			},
			inputPrimaryUDNConfig: &ovncnitypes.NetConf{
				NetConf:  cnitypes.NetConf{Name: networkName},
				Topology: ovntypes.Layer3Topology,
				NADName:  GetNADName(namespaceName, attachmentName),
				Role:     ovntypes.NetworkRolePrimary,
			},
			inputPodAnnotations: map[string]string{
				nadv1.NetworkAttachmentAnnot: GetNADName(namespaceName, "another-network"),
				DefNetworkAnnotation:         `[{"namespace": "ovn-kubernetes", "name": "default", "ips": ["192.168.0.3/24", "fda6::3/48"], "mac": "aa:bb:cc:dd:ee:ff"}]`,
			},
			expectedIsAttachmentRequested: true,
			expectedNetworkSelectionElements: map[string]*nadv1.NetworkSelectionElement{
				"ns1/attachment1": {
					Name:       "attachment1",
					Namespace:  "ns1",
					IPRequest:  nil,
					MacRequest: "",
				},
			},
			enablePreconfiguredUDNAddresses: true,
		},

```

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
Make sense updating...

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -330,13 +352,11 @@ func allocatePodAnnotationWithRollback(
 		hasIPAMClaim = ipamClaim != nil && len(ipamClaim.Status.IPs) > 0
 	}
+
 	if hasIPAM && hasStaticIPRequest {
```
That's about right, resolving the thread

qinqon commented:
Addresses review from @tssurya except possible new unit test for false negative
[force-push](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/e1f2300f0a512934e7946479491f0ae57da570c6..d8f25425462ba6adeb5acb516fb0d56c67831136)

qinqon commented:
[Going back to cni 1.2.3](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/8e7acc31990d3707a8c3fb60ad84e8391e5dfdca..a3de8680a5a8f599a34abd710938720993c2774a)

qinqon commented:
/retest

It's not related

```golang
  Unexpected error:
      <*errors.errorString | 0xc0007a3260>: 
      failed to run [killall iperf3] at VMI worker-fgzp2, error: killall iperf3 failed
      {
          s: "failed to run [killall iperf3] at VMI worker-fgzp2, error: killall iperf3 failed",
      }
  occurred
  In [It] at: /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/kubevirt.go:2320 @ 08/01/25 12:42:40.725
```

tssurya commented:
BGP CI failure looks like known flake: https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/16673702405/job/47197338115?pr=5321 - https://github.com/ovn-kubernetes/ovn-kubernetes/issues/4221

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -226,6 +226,13 @@ func validateStaticIPRequest(netInfo util.NetInfo, podDesc string) error {
+	if ipamClaim != nil && len(ipamClaim.Status.IPs) > 0 {
+		for _, ipRequest := range network.IPRequest {
+			if !util.IsItemInSlice(ipamClaim.Status.IPs, ipRequest) {
+				return fmt.Errorf("cannot allocate a static IP request with IPAM for pod %q: the pod references an ipam claim with IPs not containing the requested IP %q", podDesc, ipRequest)
```
Let's add a card and create a follow up PR then, so we can continue with this feature.

kyrtapz reviewed (approved):
LGTM

qinqon commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -226,6 +226,13 @@ func validateStaticIPRequest(netInfo util.NetInfo, podDesc string) error {
 		return fmt.Errorf("cannot allocate a static IP request with IPAM for pod %s: layer2 topology is required, but network has topology %q", podDesc, netInfo.TopologyType())
 	}
+	if ipamClaim != nil && len(ipamClaim.Status.IPs) > 0 {
+		for _, ipRequest := range network.IPRequest {
```
If we are hear we have ipRequest as stated by "validateIPRequest" so network cannot be nil.

tssurya replied:
ack sounds good, thanks for checking this.

qinqon commented:
@coderabbitai open an issue to track the error at "kv-live-migration" job, stating the "killall iperf3" error

tssurya commented:
The virt CI issue is not related to this PR: `failed to run [killall iperf3] at VMI worker-fgzp2, error: killall iperf3 failed` -> that was also seen on another PR and we have opened a new flake for it.
